File: 0022-generate-parentheses/0022-generate-parentheses.java
class Solution {
    void Func(int n, String s, List<String>res, int open, int close)
    {
        if(s.length() == 2*n)
        {
            res.add(s);
            return;
        }
        if(open < n)
            Func(n, s+"(", res,  open+1, close);
        if(close < open)
            Func(n, s+")", res,  open, close+1);
    }
    public List<String> generateParenthesis(int n) {
        List<String> res = new ArrayList<>();
        Func(n, "", res, 0, 0);
        return res;
    }
}

File: 0022-generate-parentheses/0022-generate-parentheses.cpp
class Solution {
public:
    void Func(int n, string s, vector<string>& res, int open, int close)
    {
        if(s.length() == 2*n)
        {
            res.push_back(s);
            return;
        }
        if(open < n)
            Func(n, s+ "(", res, open+1, close);
        if(close < open)
            Func(n, s+")", res, open,close+1);
    }
    vector<string> generateParenthesis(int n) {
        vector<string> res;
        Func(n,"", res,0,0);
        return res;
    }
};